test/e2e/probing_test.go

```diff
@@ -0,0 +1,393 @@
+			}
+		})
+	}
+}
+
+func hasHTTPS(servers []v1alpha3.Server) bool {
+	for _, server := range servers {
+		if server.Port.Protocol == v1alpha3.ProtocolHTTPS {
+			return true
+		}
+	}
+	return false
+}
+
+// setupHTTPS creates a self-signed certificate, installs it as a Secret and returns an *http.Transport
+// trusting the certificate as a root CA.
+func setupHTTPS(t *testing.T, hosts []string) *http.Transport {
+	t.Helper()
+
+	cert, key, err := generateCertificate(hosts)
+	if err != nil {
+		t.Fatalf("Failed to generate the certificate: %v", err)
+	}
+
+	rootCAs, _ := x509.SystemCertPool()
+	if rootCAs == nil {
+		rootCAs = x509.NewCertPool()
+	}
+
+	if ok := rootCAs.AppendCertsFromPEM(cert); !ok {
+		log.Fatalf("Failed to add the certificate to the root CA")
+	}
+
+	clients := Setup(t)
+
+	clients.KubeClient.Kube.CoreV1().Secrets("istio-system").Delete("istio-ingressgateway-certs", &metav1.DeleteOptions{})
+	_, err = clients.KubeClient.Kube.CoreV1().Secrets("istio-system").Create(&corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "istio-system",
+			Name:      "istio-ingressgateway-certs",
+		},
+		Type: corev1.SecretTypeTLS,
+		Data: map[string][]byte{
+			"tls.key": key,
+			"tls.crt": cert,
+		},
+	})
+	if err != nil {
+		t.Fatalf("Failed to set Secret %s/%s: %v", "istio-system", "istio-ingressgateway-certs", err)
+	}
+
+	return &http.Transport{
+		TLSClientConfig: &tls.Config{
+			RootCAs: rootCAs,
+		},
+	}
+}
+
+// generateCertificate generates a self-signed certificate for the provided hosts and returns
+// the PEM encoded certificate and private key.
+func generateCertificate(hosts []string) ([]byte, []byte, error) {
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate private key: %v", err)
+	}
+
+	notBefore := time.Now().Add(-5 * time.Minute)
+	notAfter := notBefore.Add(2 * time.Hour)
+
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate serial number: %v", err)
+	}
+
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			Organization: []string{"Knative Serving"},
+		},
+		NotBefore: notBefore,
+		NotAfter:  notAfter,
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+
+	for _, h := range hosts {
+		if ip := net.ParseIP(h); ip != nil {
+			template.IPAddresses = append(template.IPAddresses, ip)
+		} else {
+			template.DNSNames = append(template.DNSNames, h)
+		}
+	}
+
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, &priv.PublicKey, priv)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create the certificate: %v", err)
+	}
+
+	var certBuf bytes.Buffer
+	if err != nil {
```

Comment:
Did you mean to make a call before checking the err here?

Reply:
Good catch. This is some remaining code from when it was using files instead of bytes.Buffer.